resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -100,11 +97,56 @@ protected DnsNameResolverContext(DnsNameResolver parent,
         allowedQueries = maxAllowedQueries;
     }
 
-    protected Promise<T> promise() {
-        return promise;
+    void resolve(Promise<T> promise) {
+        boolean directSearch = parent.searchDomains().length == 0 || StringUtil.endsWith(hostname, '.');
+        if (directSearch) {
+            internalResolve(promise);
+        } else {
+            final Promise<T> original = promise;
+            promise = parent.executor().newPromise();
+            promise.addListener(new FutureListener<T>() {
+                @Override
+                public void operationComplete(Future<T> future) throws Exception {
+                    if (future.isSuccess()) {
+                        original.setSuccess(future.getNow());
+                    } else {
+                        future.addListener(new FutureListener<T>() {
+                            int count;
+                            @Override
+                            public void operationComplete(Future<T> future) throws Exception {
+                                if (future.isSuccess()) {
+                                    original.trySuccess(future.getNow());
+                                } else {
+                                    if (count < parent.searchDomains().length) {
+                                        String searchDomain = parent.searchDomains()[count++];
+                                        Promise<T> nextPromise = parent.executor().newPromise();
+                                        String nextHostname = DnsNameResolverContext.this.hostname + "." + searchDomain;
+                                        DnsNameResolverContext<T> nextContext = newResolverContext(parent,
+                                            nextHostname, resolveCache);
+                                        nextContext.internalResolve(nextPromise);
+                                        nextPromise.addListener(this);
+                                    } else {
+                                        original.tryFailure(future.cause());
+                                    }
+                                }
+                            }
+                        });
+                    }
+                }
+            });
+            int idx = hostname.length();
+            int dots = 0;
+            while (idx-- > 0) {
+                if (hostname.charAt(idx) == '.' && ++dots >= parent.ndots()) {
+                    internalResolve(promise);
```

Comment:
@Scottmitch that won't work because the listener code must be triggered whatsoever, i.e the search path must be tried in all case, but we should first try to do an internal resolve if the number of dots of the hostname is >= ndots

two examples of what the actual code does:

1/ _foo.com_ is a search path and we resolve _host1_ then we should try _host1.foo.com_ because _host1_ does not contain any dot

2/ _foo.com_ is a search path and we resolve _host2.sub_ then the following attempts should be made in this order:
- host2.sub
- host2.sub.foo.com

because _host2.sub_ has a dot in it